### PR TITLE
Add a oneshot build and allow building on a branch

### PR DESCRIPTION
**Pull request by powderluv**
Allow oneshot to build from Nod.ai's fork too.

Also remove hardcode for "main" branch

TEST=Able to build on custom shark branch
https://github.com/NodLabs/SHARK/actions/runs/2080070703

### CODE REVIEW

File: .github/workflows/oneshot_candidate_release.yml
Comment (ScottTodd) [inline .github/workflows/oneshot_candidate_release.yml:4]: This should use a different name ("Create manual candidate release" or "Oneshot candidate release" maybe)

File: .github/workflows/oneshot_candidate_release.yml
Comment (ScottTodd) [inline .github/workflows/oneshot_candidate_release.yml:11]: No need for this check with only a `workflow_dispatch` trigger.

In fact, we might be able to use the same workflow and just rework this check to something like `if: repo is google/iree || triggered by workflow dispatch`

GitHub Actions are weird, but we shouldn't have a list of "forked approved for running this workflow". Workflows can be disabled per repository, but we added this condition to avoid superflous builds in forks where there is no need to generate release artifacts on such a schedule (and those builds were failing without a write access token set IIRC).

File: .github/workflows/oneshot_candidate_release.yml
Comment (ScottTodd) [inline .github/workflows/oneshot_candidate_release.yml:50]: We do have some plans to migrate these workflows from GitHub Actions to Buildkite (first we'll migrate our presubmit/postsubmit pipelines, then release workflows, so it'll be at least a few months before much changes here specifically I think).

I'd be curious to hear more about your use case to understand why using the workflows from this repository in your downstream fork is needed.

For example, _testing_ the release pipeline from a fork for contributions back to the upstream repo is a case we want to support, but regularly generating the full set of python wheels with forked code and releasing them would raise some questions :)

**Review by ScottTodd — CHANGES_REQUESTED** [review]

File: .github/workflows/oneshot_candidate_release.yml
Comment (ScottTodd) [reply .github/workflows/oneshot_candidate_release.yml:50]: (Downstream releases are fine, but there are things like renaming "iree" to "shark" or "iree with experimental features" that I don't see the upstream workflows as being set up for - at that point creating a new release pipeline that builds what you want how you want seems better)

**Comment by stellaraccident** [comment]
(we need to rework all of this to be a simple script, and as part of that it should be relatively easy to set it up so that package names are "vendored" -- all of the legwork for that has been done but not exposed. Having a workflow to help downstreams do ad hoc releases in the meantime is fine with me. Just please don't add things to pypi that borrow our package names as that invites brokenness in the field)

File: .github/workflows/oneshot_candidate_release.yml
Comment (powderluv) [reply .github/workflows/oneshot_candidate_release.yml:4]: sounds good. will do.

Comment (powderluv) [reply .github/workflows/oneshot_candidate_release.yml:4]: done

File: .github/workflows/oneshot_candidate_release.yml
Comment (powderluv) [reply .github/workflows/oneshot_candidate_release.yml:11]: ah makes sense. Will remove.

File: .github/workflows/oneshot_candidate_release.yml
Comment (powderluv) [reply .github/workflows/oneshot_candidate_release.yml:50]: ideally we don't want any downstream fork of IREE. The only thing we are carry indefinitely are the CUDA caching allocator but hopefully that will also morph into something else upstream. I plan to use a different version tag (the -suffix) seems to work. 

Overall we want to provide a way to replace stock IREE with this version via pip if possible. If we have to create a new package name and namespace then we have to create a "deep fork" to rename the namespace etc and we would rather not. 

We want downstream users (and other customers to be able to install by pointing to something like https://github.com/NodLabs/SHARK/releases/tag/untagged-bb5737a2d500d03df898 )

**Comment by powderluv** [comment]
> (we need to rework all of this to be a simple script, and as part of that it should be relatively easy to set it up so that package names are "vendored" -- all of the legwork for that has been done but not exposed. Having a workflow to help downstreams do ad hoc releases in the meantime is fine with me. Just please don't add things to pypi that borrow our package names as that invites brokenness in the field)

yes yes. we don't plan to add any version of IREE to PyPI.  We will have our "shark_runner" which has all our orchestration code that eventually would be in pypi.

File: .github/workflows/oneshot_candidate_release.yml
Comment (powderluv) [reply .github/workflows/oneshot_candidate_release.yml:50]: also it may be helpful for customers building downstream vendorized versions if we keep the release in github actions.

**Comment by powderluv** [comment]
@ScottTodd PTAL

**Review by ScottTodd — APPROVED** [review]

**Comment by ScottTodd** [comment]
> @ScottTodd PTAL

Ah, missed this, thanks for the ping. The "re-request review" button gets a PR to show up in the "awaiting review from you" filter:
image

**Pull request merged**
